### ISSUE TITLE
FEATURE REQUEST - Make metrics accessible outside the CGI

### ISSUE THREAD

**Issue by unixorn** [moosefs/moosefs#290]
## Have you read through available documentation and open Github issues?

Yes

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?

FEATURE request

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).

3.0.103

#### Operating system (distribution) and kernel version.

Mix of debian stretch and buster.

I would like to have a way to extract the metrics that are currently displayed in the CGI and write them to `statsd` or `prometheus`.

That would make it much easier to set up proper alerting rules for various cluster conditions that I can currently only see in the CGI interface.

**Comment by acid-maker** [comment]
@xandrus is working on JSON output to mfscli - I hope that it will be enough.

**Comment by unixorn** [comment]
That would be great. If I can run `something --json > state.json`, I can parse the state.json and write the metrics I want to `statsd`, `prometheus` or wherever.

**Comment by xandrus** [comment]
Hi,
@acid-maker thanks, that you assigned this task to me.
Basically, I have started adding JSON output to mfscgi command a few weeks ago but at the moment I have focused on a different task. 
Personally I believe that this week I will be able to spend more time on this task.

@unixorn I would like to ask you which mfscli data set would you like to have first in JSON output.
I mean
```
	-S data set : defines data set to be displayed
		-SIN : show full master info
		-SIM : show only masters states
		-SLI : show only licence info
		-SIG : show only general master (leader) info
		-SMU : show only master memory usage
		-SIC : show only chunks info (goal/copies matrices)
		-SIL : show only loop info (with messages)
		-SMF : show only missing chunks/files
		-SCS : show connected chunk servers
		-SMB : show connected metadata backup servers
		-SHD : show hdd data
		-SEX : show exports
		-SMS : show active mounts
		-SRS : show resources (storage classes,open files,acquired locks)
		-SSC : show storage classes
		-SOF : show only open files
		-SAL : show only acquired locks
		-SMO : show operation counters
		-SQU : show quota info
		-SMC : show master charts data
		-SCC : show chunkserver charts data
```

**Comment by oszafraniec** [comment]
@xandrus , I think I need to come back to our abandoned project we have started last year "after hours" ;) I think we can make decent use of your json output instead of parsing mfscli output we used to do... What do you think? :) 

And of course our work was inspired by [Ceph Dashboard](https://github.com/Crapworks/ceph-dash), there is no doubt about it.

![IMG_9642](https://user-images.githubusercontent.com/3938516/68799973-81566400-0659-11ea-8681-bda65e127e28.JPG)

![Zrzut ekranu 2018-07-25 o 12 58 12](https://user-images.githubusercontent.com/3938516/68799974-81566400-0659-11ea-974a-096b8a424357.png)

**Comment by jkiebzak** [comment]
+1

>

**Comment by unixorn** [comment]
@xandrus, the first things I'd like to see are the connected server counts (both chunk servers and metadata servers), missing chunk data, and hdd data.

**Comment by xandrus** [comment]
@oszafraniec, all the most important information on one page - nice!
@unixorn, so I will focus on these resources. 

Thanks.

**Comment by s5unty** [comment]
[parse the mfscli output and write the metrics](https://gist.github.com/s5unty/e6eace30107fa548e2320a5dcfbe5695)

**Comment by tianon** [comment]
@s5unty I started something similar :smile:

https://gist.github.com/tianon/8db9b120fe83b7ed24c30b7911d739c9

It'd be awesome if we could get something slightly more official (especially standardized metric names). :innocent:

**Comment by s5unty** [comment]
AWK is nice, but [google/mtail](https://github.com/google/mtail) is more fun. :laughing:  [WIP](https://gist.github.com/s5unty/fc39e92efabac15c41a8c94724cf1039)

**Comment by tianon** [comment]
Looking at the `mfscli` code, scraping this output can't possibly scale -- there are a lot of instances of `if masterconn.version_at_least(x,y,z): ... data.append(...)` (in other words, columns that only even show up in the data at all if the master is over a certain version), so which column corresponds to which data is dependent on the master version number. :weary:

**Comment by uu** [comment]
@tianon thank you for the gist! Do you have any grafana dashboard for it?

**Comment by eleaner** [comment]
I see we still don't have a proper exporter. That's a shame. The extension of cgiserv should be a piece of cake.

could someone point me - on how to use the awk file (one or the other attached above) to get the data into prometheus?

**Comment by maxlivi** [comment]
I am using grafan/influxdb/telegraf.
I made i simple script for the tha feed that database ( not using the last flux 2.x).
The commands are very simple, like:

...
mfscli -SIG|awk '/store/{print "moosefs,metric=last_metadata_sync lstupdt=""\""strftime("%d %b %T %Z",$6)"\""}'
mfscli -SIC -p | awk '/stable/ {print "moosefs,metric=chunks_stable stable="$6"i"}'
mfscli -SIC -p | awk '/undergoal:/ {print "moosefs,metric=chunks_under_goal undergoal="$6"i"}'
mfscli -SIC -p | awk '/endangered:/{print "moosefs,metric=chunks_endangered endangered="$6"i"}
...

Then I left ncat listening on an arbitrary port and when connect is launch the script.

Telegraf has the following to connect to ncat:

 [[inputs.http]]
#   ## One or more URLs from which to read formatted metrics
   urls = [
	"http://localhost:9473", "http://localhost:9474"
   ]

Below a screenshot of  grafana dashboard that has the alerts set for the most important events

![2022-08-25-114330_1285x912_scrot](https://user-images.githubusercontent.com/60069365/186646353-82404a96-42b3-4933-9c97-7c350542cd67.png)

**Comment by eleaner** [comment]
that's pretty neat
I see everybody is using grafana/influxdb/telegraf

I was using Zabbix for years (I mean, I configured it like 5 years ago, and the nit was running), but LonghornFS below my db died on me recently, and I could not figure out how to revive it.
So I started playing with grafana and prometheus, as you can imagine I am pretty new to all that

Could you tell me how is grafana/influxdb/telegraf superior to grafana/prometheus?

**Comment by maxlivi** [comment]
I don't think that grafana/influxdb/telegraf is superior to grafana/prometheus.

Prometheus is "a systems and service monitoring system. It collects metrics from configured targets at given intervals, evaluates rule expressions, displays the results, and can trigger alerts if some condition is observed to be true." 

Telegraf is just a collector, it is a single small go binary with a minimum memory footprint, it has just one configuration file and can collect many, many metrics.

It can also collect metrics from the sysstats tools.

I know very little of Prometheus having used Telegraf/Influx/Grafana all the time.

I am also running zabbix in a lxd container and I like the possibility to automatically start a service if it crashes.

Zabbix has also nice infrastructure visualisation.

**Comment by seecsea** [comment]
> I am using grafan/influxdb/telegraf. I made a simple script that feeds the database ( not using the last flux 2.x). The commands are very simple, like:
> 
> ```
> ...
> mfscli -SIG|awk '/store/{print "moosefs,metric=last_metadata_sync lstupdt=""\""strftime("%d %b %T %Z",$6)"\""}'
> mfscli -SIC -p | awk '/stable/ {print "moosefs,metric=chunks_stable stable="$6"i"}'
> mfscli -SIC -p | awk '/undergoal:/ {print "moosefs,metric=chunks_under_goal undergoal="$6"i"}'
> mfscli -SIC -p | awk '/endangered:/{print "moosefs,metric=chunks_endangered endangered="$6"i"}
> ...
> ```
> 
> Then I left ncat listening on an arbitrary port and when connected it launches the script.
> 
> Telegraf has the following to connect to ncat:
> 
> ```
>  [[inputs.http]]
>    urls = [
> 	"http://localhost:9473", "http://localhost:9474"
>    ]
> ```
> 
> Below a screenshot of grafana dashboard that has the alerts set for the most important events
> 
> ![2022-08-25-114330_1285x912_scrot](https://user-images.githubusercontent.com/60069365/186646353-82404a96-42b3-4933-9c97-7c350542cd67.png)

Hi, guy. I am using grafan/influxdb/telegraf too.
Can you share the commands and the dashboard json? 

https://grafana.com/grafana/dashboards/16700-moosefs-overview/
There is a prometheus dashborad, but I do not use Prometheus yet.

**Comment by maxlivi** [comment]
Hello eecsea,

no problems on sharing my script and the dashboard but I have switched the DB from influxdb to postgres+timescale so the dashboard is not going to work if you are using influx1/2.

If you are still interested I will upload everything on my site or upload here if it is  permitted

Cheers

**Comment by tianon** [comment]
I finally got around to playing with the new `mfscli -j` option in the new 4.x release and wrote a ton of `jq` to have something that still needs a lot of work (and I'm very much not sold on the metric names), but at least is going to be a lot more reliable/stable than my previous awk-based solution scraping the columnar output of `mfscli`: https://github.com/tianon/node-exporter-textfile-scripts/blob/7e23f5835684c4ee7849a7a7fa9f665d7d317a2e/mfs.jq :eyes:

(I'm sure it's not recommended, but I also had luck running/testing this with `mfscli` version 4+ against my existing version 3 cluster :sweat_smile:)

**Comment by chogata** [comment]
Indeed :) So, closing :)
